Documents API should handle `OverloadedException` gracefully

(note: while this issue is specifically about Docs API, same/similar issues are very likely with REST and GraphQL APIs -- we'll file separate issues for those, as necessary)

Looking at our logs we noticed that there are error entries with stack trace, correlating to spikes in 500 responses for Documents API end points. Stack traces look like:

```
     Error when executing request
     java.util.concurrent.CompletionException: org.apache.cassandra.stargate.exceptions.OverloadedException: Rate limit reached, please consult our documentation at https://astra.dev/rate-limit
     	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
     	at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:670)
     	at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:658)
     	at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2094)
     	at io.stargate.db.datastore.PersistenceBackedDataStore.batch(PersistenceBackedDataStore.java:201)
     	at io.stargate.db.datastore.PersistenceBackedDataStore.batch(PersistenceBackedDataStore.java:190)
     	at io.stargate.db.query.AsyncQueryExecutor.batch(AsyncQueryExecutor.java:104)
     	at io.stargate.db.query.AsyncQueryExecutor.batch(AsyncQueryExecutor.java:82)
     	at io.stargate.web.docsapi.service.write.DocumentWriteService.executeBatch(DocumentWriteService.java:498)
     	at io.stargate.web.docsapi.service.write.DocumentWriteService.lambda$writeDocument$3(DocumentWriteService.java:125)
     	at io.reactivex.rxjava3.internal.operators.single.SingleFlatMap$SingleFlatMapCallback.onSuccess(SingleFlatMap.java:77)
     	at io.reactivex.rxjava3.internal.operators.single.SingleMap$MapSingleObserver.onSuccess(SingleMap.java:65)
     	at io.reactivex.rxjava3.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run(SingleObserveOn.java:81)
     	at io.reactivex.rxjava3.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38)
     	at io.reactivex.rxjava3.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:25)
     	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
     	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
     	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
     	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
     	at java.base/java.lang.Thread.run(Thread.java:834)
     Caused by: org.apache.cassandra.stargate.exceptions.OverloadedException: Rate limit reached, please consult our documentation at https://astra.dev/rate-limit
 
```

This is from the current Stargate (V1), SGv2 probably has similar handling.

Instead of this exception propagating through, we should catch it and give caller something like `429` (Too Many Requests (RFC 6585)) so that it can reduce send rate.

--- a/restapi/src/main/java/io/stargate/web/docsapi/rx/RxUtils.java
+++ b/restapi/src/main/java/io/stargate/web/docsapi/rx/RxUtils.java
@@ -19,6 +19,7 @@ package io.stargate.web.docsapi.rx;
 
 import io.reactivex.rxjava3.core.Single;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
 import java.util.function.Supplier;
 
 public final class RxUtils {
@@ -33,16 +34,24 @@ public final class RxUtils {
                 .whenComplete(
                     (e, t) -> {
                       if (null != t) {
-                        emitter.onError(t);
+                        emitter.onError(getEmittingThrowable(t));
                       } else {
                         emitter.onSuccess(e);
                       }
                     })
                 .whenComplete(
                     (e, t) -> {
-                      if (null != t) {
-                        emitter.onError(t);
+                      if (null != t && !emitter.isDisposed()) {
+                        emitter.onError(getEmittingThrowable(t));
                       }
                     }));
+  }
+
+  private static Throwable getEmittingThrowable(Throwable t) {
+    if (t instanceof CompletionException && null != t.getCause()) {
+      return t.getCause();
+    } else {
+      return t;
+    }
   }
 }

--- a/restapi/src/test/java/io/stargate/db/datastore/ValidatingDataStore.java
+++ b/restapi/src/test/java/io/stargate/db/datastore/ValidatingDataStore.java
@@ -269,6 +269,7 @@ public class ValidatingDataStore implements DataStore {
     private int pageSize = Integer.MAX_VALUE;
     private BatchType batchType;
     private List<Map<String, Object>> rows;
+    private RuntimeException executionException;
 
     private QueryExpectation(Table table, String cqlRegEx, Object[] params) {
       this.cqlPattern = Pattern.compile(cqlRegEx);
@@ -289,6 +290,11 @@ public class ValidatingDataStore implements DataStore {
 
     public QueryExpectation inBatch(BatchType batchType) {
       this.batchType = batchType;
+      return this;
+    }
+
+    public QueryExpectation withExecutionException(RuntimeException e) {
+      this.executionException = e;
       return this;
     }
 
@@ -314,6 +320,13 @@ public class ValidatingDataStore implements DataStore {
     }
 
     private CompletableFuture<ResultSet> execute(Parameters parameters, BatchType batchType) {
+      if (null != executionException) {
+        return CompletableFuture.supplyAsync(
+            () -> {
+              throw executionException;
+            });
+      }
+
       Optional<ByteBuffer> pagingState = parameters.pagingState();
       int pageSize;
       if (this.pageSize < Integer.MAX_VALUE) {

--- a/restapi/src/test/java/io/stargate/web/docsapi/service/QueryExecutorTest.java
+++ b/restapi/src/test/java/io/stargate/web/docsapi/service/QueryExecutorTest.java
@@ -44,6 +44,7 @@ import io.stargate.db.schema.Table;
 import java.nio.ByteBuffer;
 import java.util.List;
 import java.util.Map;
+import org.apache.cassandra.stargate.exceptions.OverloadedException;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
@@ -123,6 +124,16 @@ class QueryExecutorTest extends AbstractDataStoreTest {
 
     List<RawDocument> r1 = values(executor.queryDocs(allDocsQuery, 100, false, null, context));
     assertThat(r1).extracting(RawDocument::id).containsExactly("1", "2");
+  }
+
+  @Test
+  void testFullScanWithExecutionException() throws Exception {
+    OverloadedException error = new OverloadedException("Testing purposes");
+    withQuery(table, "SELECT * FROM %s").withExecutionException(error);
+
+    executor.queryDocs(allDocsQuery, 100, false, null, context).test().await().assertError(error);
+
+    resetExpectations();
   }
 
   private void withFiveTestDocs(int pageSize) {

